### PR TITLE
Allow qualification to be done with NOWAIT AIO support (RWF_NOWAIT)

### PR DESCRIPTION
**Pull request by raphaelsc**
```
$ sudo ~/fsqual/fsqual
memory DMA alignment:         512
disk DMA read alignment:      512
disk DMA write alignment:      512
context switch per appending io (mode size-changing, iodepth 1): 0 (GOOD)
context switch per appending io (mode size-changing, iodepth 3): 0.7932 (BAD)
context switch per appending io (mode size-unchanging, iodepth 3): 0.0002 (GOOD)
context switch per appending io (mode size-unchanging, iodepth 7): 0.0001 (GOOD)

$ sudo ~/fsqual/fsqual --nowait
memory DMA alignment:         512
disk DMA read alignment:      512
disk DMA write alignment:      512
context switch per appending io (mode size-changing, iodepth 1): 0 (GOOD)
context switch per appending io (mode size-changing, iodepth 3): 0 (GOOD)
context switch per appending io (mode size-unchanging, iodepth 3): 0 (GOOD)
context switch per appending io (mode size-unchanging, iodepth 7): 0 (GOOD)

$ sudo ~/fsqual/fsqual --help
Usage: /home/raphaelsc/fsqual/fsqual <option(s)>
Options:
	-h,--help		Show this help message
	--nowait	Use NOWAIT AIO (RWF_NOWAIT flag)

Signed-off-by: Raphael S. Carvalho <raphaelsc@scylladb.com>
```

### CODE REVIEW

File: fsqual.cc [hunk 1]
```diff
@@ -57,7 +57,11 @@ void test_concurrent_append(io_context_t ioctx, int fd, unsigned iodepth, std::s
     while (completed < nr) {
         auto i = unsigned(0);
         while (initiated < nr && current_depth < iodepth) {
-            io_prep_pwrite(&iocbs[i++], fd, buf, bufsize, bufsize*initiated++);
+            io_prep_pwrite(&iocbs[i], fd, buf, bufsize, bufsize*initiated++);
+            if (nowait) {
+                iocbs[i].aio_rw_flags |= RWF_NOWAIT;
```
Comment (avikivity) [inline fsqual.cc:62]: Need to protect against RWF_NOWAIT not being defined on old systems.

File: fsqual.cc [hunk 2]
```diff
@@ -120,11 +124,31 @@ void test_dio_info() {
...
+
+    for (int i = 1; i < ac; ++i) {
+        std::string arg = av[i];
+        if ((arg == "-h") || (arg == "--help")) {
+            show_usage(av[0]);
+            return 0;
+        } else if (arg == "--nowait") {
+            nowait = true;
+        }
+    }
```
Comment (avikivity) [inline fsqual.cc:146]: Please use a real argument parser.

File: fsqual.cc (hunk 1 above)
Comment (avikivity) [reply fsqual.cc:62]: Shouldn't we handle failures and resubmit them? Or at least count and print them?

**Comment by raphaelsc** [comment]
On Wed, Jul 22, 2020 at 12:50 PM Avi Kivity <notifications@github.com>
wrote:

> *@avikivity* commented on this pull request.
> ------------------------------
>
> In fsqual.cc
> <https://github.com/avikivity/fsqual/pull/2#discussion_r458892991>:
>
> > @@ -57,7 +57,11 @@ void test_concurrent_append(io_context_t ioctx, int fd, unsigned iodepth, std::s
>      while (completed < nr) {
>          auto i = unsigned(0);
>          while (initiated < nr && current_depth < iodepth) {
> -            io_prep_pwrite(&iocbs[i++], fd, buf, bufsize, bufsize*initiated++);
> +            io_prep_pwrite(&iocbs[i], fd, buf, bufsize, bufsize*initiated++);
> +            if (nowait) {
> +                iocbs[i].aio_rw_flags |= RWF_NOWAIT;
>
> Need to protect against RWF_NOWAIT not being defined on old systems.
>
Do you mind if I detect it after calling io_submit() to facilitate my life?

Something like:
ret = io_submit(...);
if (nowait && ret == EINVAL) { exit(bla bla bla) }




> ------------------------------
>
> In fsqual.cc
> <https://github.com/avikivity/fsqual/pull/2#discussion_r458893872>:
>
> >  int main(int ac, char** av) {
> +    bool nowait = false;
> +
> +    for (int i = 1; i < ac; ++i) {
> +        std::string arg = av[i];
> +        if ((arg == "-h") || (arg == "--help")) {
> +            show_usage(av[0]);
> +            return 0;
> +        } else if (arg == "--nowait") {
> +            nowait = true;
> +        }
> +    }
>
> Please use a real argument parser.
>
Could use boost's program options, but I came up with my own parser to
avoid introducing a library dependency. Do you think it's not a problem?


> ------------------------------
>
> In fsqual.cc
> <https://github.com/avikivity/fsqual/pull/2#discussion_r458895285>:
>
> > @@ -57,7 +57,11 @@ void test_concurrent_append(io_context_t ioctx, int fd, unsigned iodepth, std::s
>      while (completed < nr) {
>          auto i = unsigned(0);
>          while (initiated < nr && current_depth < iodepth) {
> -            io_prep_pwrite(&iocbs[i++], fd, buf, bufsize, bufsize*initiated++);
> +            io_prep_pwrite(&iocbs[i], fd, buf, bufsize, bufsize*initiated++);
> +            if (nowait) {
> +                iocbs[i].aio_rw_flags |= RWF_NOWAIT;
>
> Shouldn't we handle failures and resubmit them? Or at least count and
> print them?
>
You're right. We need to make sure requests that failed with EAGAIN are
resubmitted so that we complete all requests.

**Comment by avikivity** [comment]
> On Wed, Jul 22, 2020 at 12:50 PM Avi Kivity ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In fsqual.cc <[#2 (comment)](https://github.com/avikivity/fsqual/pull/2#discussion_r458892991)>: > @@ -57,7 +57,11 @@ void test_concurrent_append(io_context_t ioctx, int fd, unsigned iodepth, std::s while (completed < nr) { auto i = unsigned(0); while (initiated < nr && current_depth < iodepth) { - io_prep_pwrite(&iocbs[i++], fd, buf, bufsize, bufsize*initiated++); + io_prep_pwrite(&iocbs[i], fd, buf, bufsize, bufsize*initiated++); + if (nowait) { + iocbs[i].aio_rw_flags |= RWF_NOWAIT; Need to protect against RWF_NOWAIT not being defined on old systems.
> Do you mind if I detect it after calling io_submit() to facilitate my life? Something like: ret = io_submit(...); if (nowait && ret == EINVAL) { exit(bla bla bla) }

If it's not defined, the program will not compile.

> ------------------------------ In fsqual.cc <[#2 (comment)](https://github.com/avikivity/fsqual/pull/2#discussion_r458893872)>: > int main(int ac, char** av) { + bool nowait = false; + + for (int i = 1; i < ac; ++i) { + std::string arg = av[i]; + if ((arg == "-h") || (arg == "--help")) { + show_usage(av[0]); + return 0; + } else if (arg == "--nowait") { + nowait = true; + } + } Please use a real argument parser.
> Could use boost's program options, but I came up with my own parser to avoid introducing a library dependency. Do you think it's not a problem?

It's better to do it right.

**Pull request closed without merging**